contracts/EtomicSwapNft.sol

```diff
@@ -0,0 +1,719 @@
+    event TakerPaymentRefundedTimelock(bytes32 id);
+
+    mapping(bytes32 => TakerPayment) public takerPayments;
+
+    address public immutable dexFeeAddress;
+
+    constructor(address feeAddress) {
+        require(
+            feeAddress != address(0),
+            "feeAddress must not be zero address"
+        );
+
+        dexFeeAddress = feeAddress;
+    }
+
+    function spendErc721MakerPayment(
+        bytes32 id,
+        address maker,
+        bytes32 takerSecretHash,
+        bytes32 makerSecret,
+        address tokenAddress,
+        uint256 tokenId
+    ) external {
+        require(
+            makerPayments[id].state == MakerPaymentState.PaymentSent,
+            "Invalid payment state. Must be PaymentSent"
+        );
+
+        // Check if the function caller is an externally owned account (EOA)
+        require(msg.sender == tx.origin, "Caller must be an EOA");
+
+        bytes20 paymentHash = ripemd160(
+            abi.encodePacked(
+                msg.sender,
+                maker,
+                takerSecretHash,
+                sha256(abi.encodePacked(makerSecret)),
+                tokenAddress,
+                tokenId
+            )
+        );
+        require(
+            paymentHash == makerPayments[id].paymentHash,
+            "Invalid paymentHash"
+        );
+
+        // Effects
+        makerPayments[id].state = MakerPaymentState.TakerSpent;
+
+        // Event Emission
+        emit MakerPaymentSpent(id);
+
+        // Interactions
+        IERC721 token = IERC721(tokenAddress);
+        token.safeTransferFrom(address(this), msg.sender, tokenId);
+    }
+
+    function spendErc1155MakerPayment(
+        bytes32 id,
+        uint256 amount,
+        address maker,
+        bytes32 takerSecretHash,
+        bytes32 makerSecret,
+        address tokenAddress,
+        uint256 tokenId
+    ) external {
+        require(
+            makerPayments[id].state == MakerPaymentState.PaymentSent,
+            "Invalid payment state. Must be PaymentSent"
+        );
+
+        // Check if the function caller is an externally owned account (EOA)
+        require(msg.sender == tx.origin, "Caller must be an EOA");
+
+        bytes20 paymentHash = ripemd160(
+            abi.encodePacked(
+                msg.sender,
+                maker,
+                takerSecretHash,
+                sha256(abi.encodePacked(makerSecret)),
+                tokenAddress,
+                tokenId,
+                amount
+            )
+        );
+        require(
+            paymentHash == makerPayments[id].paymentHash,
+            "Invalid paymentHash"
+        );
+
+        // Effects
+        makerPayments[id].state = MakerPaymentState.TakerSpent;
+
+        // Event Emission
+        emit MakerPaymentSpent(id);
+
+        // Interactions
+        IERC1155 token = IERC1155(tokenAddress);
+        token.safeTransferFrom(address(this), msg.sender, tokenId, amount, "");
```

Comment:
ooh, I see that you followed the same order as `safeTransferFrom` where you put amount at the end. I think the calling from mm2 should be the same order for ERC20 or NFTs though.

Reply:
> you followed the same order as safeTransferFrom where you put amount at the end.

yep

> I think the calling from mm2 should be the same order for ERC20 or NFTs though.

Oh, yes, I forgot to change amount position in erc1155 functions fields. Will move `uint256 amount` below `tokenId`

as for rust structures for SwapOps args, I followed the same order as for coin args, but with additional nft fields in the end. <- did it in the case if we decide to re utilize same args structs for coin and nft swaps. (But I think )

https://github.com/KomodoPlatform/komodo-defi-framework/blob/adfe045ccb6613ed039a381a120d89f9338b7742/mm2src/coins/lp_coins.rs#L1545-L1566

```rust
    /// The address of the NFT token
    pub token_address: &'a [u8],
    /// The ID of the NFT token.
    pub token_id: &'a [u8],
    /// The type of smart contract that governs this NFT
    pub contract_type: &'a [u8],
    // Etomic swap contract address
    pub swap_contract_address: &'a [u8],
```

I will change amount order for `EtomicSwapNft` erc1155 functions. 
As for rust structs `args` I suggest to leave them as they are, We can change them during refactoring/changes in the future (during full nft trading proto V2 support development)